library_generation/test/generate_library_unit_tests.sh

```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env bash
+
+set -xeo pipefail
+
+# Variables used to generate final result
+total_num=0
+succeed_num=0
+failed_num=0
+# Unit tests against ./utilities.sh
+script_dir=$(dirname "$(readlink -f "$0")")
+source "$script_dir"/../utilities.sh
+
+assertEquals() {
```

Comment:
We can move these helper methods to the bottom and/or add underscore prefix to indicate they should only be used internally, see this [StackOverflow post](https://stackoverflow.com/questions/2289293/is-there-a-convention-for-naming-private-functions-in-bash).

Reply:
I renamed the helper functions.

However, is it a convention to move helper functions to the bottom?

Reply:
> However, is it a convention to move helper functions to the bottom?

I don't think there is any written convention regarding this. From my personal experience and preference, making helpers the first thing in the tests may take away people's attention from real unit tests, especially if you don't have a good naming convention for unit tests. I see that it's much better now with proper names for the helper functions and unit tests, so I'm OK with the way it is.